### PR TITLE
Replace OCPWin and OCaml with esy

### PR DESCRIPTION
**Pull request by andschwa**
This replaces the CMake-driven OCaml build with [`esy`](https://esy.sh/) and `dune`, which means that our dependencies drop down to just installing `esy` via `npm`, and it builds a native `oeedger8r` executable across all our platforms.

Rough edges:
- [x] It builds in the source tree instead of the build tree, grr.
- [x] We need `npm install -g esy` done to our CI systems.
- [x] Update Ansible scripts.
- [x] Update Windows install script.
- [x] Still need to update the CI containers.
- [x] I need to setup the correct `.gitignore` file and commit the `esy.lock` directory to vendor our dependencies.

Gains: no more OCPWin! Modern OCaml! OCaml packages! Dune build! (And actually, no `npm` requirement on Linux, only on Windows, and only for now.)

### CODE REVIEW

**Comment by andschwa** [comment]
To test:

````
apt install npm
npm install -g esy
```

And then test as normal.

**Comment by andschwa** [comment]
Oh, I'd also like to setup auto-formatting as esy and dune make that easy!

**Comment by BRMcLaren** [comment]
Hey Andy, 

This will relate to something I am working on.

As it changes dependencies, I just wanted to check if this was slated for 0.7?

**Comment by andschwa** [comment]
@BRMcLaren Yeah, let's get this in ASAP. I don't believe the Windows work can really go forward without it.

**Comment by andschwa** [comment]
Running end to end tests [here](https://oe-jenkins.eastus.cloudapp.azure.com/blue/organizations/jenkins/OpenEnclave-E2E-Testing/detail/OpenEnclave-E2E-Testing/56/pipeline).

**Comment by andschwa** [comment]
I believe this is ready for review, though not for merging as we need all the CI systems actually updated (not just the few @johnkord and I did to run some of the tests).

**Comment by andschwa** [comment]
Ya'll Ansible is awful, the docs aren't right 😭

**Comment by andschwa** [comment]
@johnkord @radhikaj @CodeMonkeyLeet Could use a review here. Last build (which should work as all it was missing was an addition to the `PATH` on Windows) is [here](https://oe-jenkins.eastus.cloudapp.azure.com/blue/organizations/jenkins/OpenEnclave-E2E-Testing/detail/OpenEnclave-E2E-Testing/62/pipeline).

The only changes here really are to the build. The OCaml changes were just auto-formatting with the updated (and now locked) version of `ocamlformat` (and some warning fixes we can now do with the up-to-date version of OCaml).

Notably, this does _not_ add NPM as a dependency to Linux.

File: scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml [hunk 1]
```diff
@@ -30,35 +30,8 @@
...
-- name: Clean OCaml downloaded deb packages
-  file:
-    state: absent
-    path: "{{ item.value.local_path }}"
-  with_dict: "{{ ocaml_packages }}"
-
-- import_role:
-    name: linux/openenclave
-    tasks_from: opam-setup.yml
+- name: Install esy
```
Comment (johnkord) [inline scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml]: Thank goodness this is gone

File: tools/oeedger8r/package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "oeedger8r",
+  "version": "0.6.0",
```
Comment (johnkord) [inline tools/oeedger8r/package.json]: If we update our version number (which we will do to 0.7.0 soon) I'd like this to not get left behind. Maybe we should make a comment in the RELEASING.md document that tells the release manager to update this file?

File: tools/oeedger8r/package.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      "includePackages": [
+        "root"
+      ]
+    }
+  },
+
+  "dependencies": {
+    "@opam/dune": "~1.11.3",
+    "ocaml": "~4.6.0"
+  },
```
Comment (johnkord) [inline tools/oeedger8r/package.json]: If we need to update ocaml (or dune/merlin/ocamlformat), we can update the version here in this file.

File: scripts/ansible/oe-windows-acc-setup.yml [hunk 4]
```diff
@@ -20,8 +20,8 @@
         seven_zip_hash:      "F00E1588ED54DDF633D8652EB89D0A8F95BD80CCCFC3EED362D81927BEC05AA5"
         vs_buildtools:       "https://oejenkins.blob.core.windows.net/oejenkins/vs_buildtools_2017.exe"
         vs_buildtools_hash:  "6F49872B04A0EAEDF5ED96AB25F7697062D81A419D45D9970E41784F31165BF2"
-        ocaml_url:           "https://oejenkins.blob.core.windows.net/oejenkins/ocpwin64-20160113-4.02.1+ocp1-mingw64.zip"
-        ocaml_hash:          "369F900F7CDA543ABF674520ED6004CC75008E10BEED0D34845E8A42866D0F3A"
+        node_url:            "https://nodejs.org/dist/v10.16.3/node-v10.16.3-x64.msi"
+        node_hash:           "f68b75eea46232adb8fd38126c977dc244166d29e7c6cd2df930b460c38590a9"
```
Comment (johnkord) [inline scripts/ansible/oe-windows-acc-setup.yml]: Let's create an issue to cache this node*.msi file in the oejenkins storage account for this ansible playbook

**Comment by andschwa** [comment]
The commits touching Ansible and PowerShell were spun off into #2246.

**Comment by andschwa** [comment]
Just FYI this is on hold until we solve other problems with our Windows CI systems (the `PATH` environment variable has gotten too long 😂)

**Comment by andschwa** [comment]
Good news! We fixed the Windows machines. Meh news: we can't install `ocamlformat` (for now) as it has broken symlinks which crashes `tar` on Windows...esy folks are fixing it. Better news: we can merge this (soon) without having to replace the CI machines, and then we can just do that in one go while removing the old OCaml tooling.

**Comment by andschwa** [comment]
bors try

**Comment by andschwa** [comment]
bors try-

**Comment by andschwa** [comment]
Separated out some changes for easier reverting in the future.

**Comment by andschwa** [comment]
bors try

**Comment by andschwa** [comment]
Sorry this took so long folks, a lot of the CI responsibilities got shuffled around at the same time, but it should be in soon (FYI @anakrish).

File: scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml (hunk 1 above)
Comment (anakrish) [reply scripts/ansible/roles/linux/openenclave/tasks/environment-setup.yml:33]: @johnkord What were the issues with installing ocaml on Linux?

**Pull request merged**